Propagate TLS callbacks with ThreadLocalState

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* #37639 Use TensorMethods.cpp
* #37775 Memory profiling
* #37774 Move ThreadLocalDebugInfo to c10
* #36291 Make profiler thread local
* **#37745 Propagate TLS callbacks with ThreadLocalState**
* #37548 Move RecordFunction into ATen
* #37491 Make RecordFunction callbacks thread local and modernize interface

Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

Commit:
Propagate TLS callbacks with ThreadLocalState

Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/record_function.cpp:
```diff
@@ -199,6 +199,23 @@ inline CallbackManager& manager() {
+}
+
+void _setTLSCallbacks(const RecordFunctionCallbacks& callbacks) {
+  clearThreadLocalCallbacks();
```
technically you just assign it next so you don't need it?

dzhulgakov commented on aten/src/ATen/record_function.cpp:
```diff
@@ -199,6 +199,23 @@ inline CallbackManager& manager() {
+  clearThreadLocalCallbacks();
+  // keep the original handles
+  sorted_tls_callbacks_ = callbacks;
+  std::sort(
```
is it just a precaution?

ilia-cher replied:
yes, I assume this function is not perf sensitive and leaving it up to caller would be too fragile

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]

Commit:
Update on "Propagate TLS callbacks with ThreadLocalState"


Summary:
This PR makes it possible to set TLS callbacks and use
them transparently not only in the main thread but also
in any async tasks

Differential Revision: [D21374873](https://our.internmc.facebook.com/intern/diff/D21374873)

[ghstack-poisoned]
